### ISSUE TITLE
Training consistently fails.

### ISSUE THREAD

**Issue by crazyherozk** [Mcourtyard/m-courtyard#8]
Hi, help me
The model I'm using is quite small. My hardware and training configurations are shown in the attached image, but the process fails every time. Here is the error message:

libc++abi: terminating due to uncaught exception of type std::runtime_error: [METAL] Command buffer execution failed: Impacting Interactivity (0000000e:kIOGPUCommandBufferCallbackErrorImpactingInteractivity)

/opt/homebrew/Cellar/python@3.12/3.12.13/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/resource_tracker.py:279: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown

<img width="2056" height="1329" alt="Image" src="https://github.com/user-attachments/assets/72ab9d2c-c523-467e-94fc-1d66e83f405e" />

**Comment by tuwenbo0120** [comment]
Thanks for the report. This error is macOS's Metal watchdog (kIOGPUCommandBufferCallbackErrorImpactingInteractivity) — the system killed the GPU task because a single training step ran long enough to impact UI responsiveness. It's not an app crash or an MLX bug per se, but a signal the current config is too heavy for your GPU in that moment.

Please try:

Batch size → 1
Max sequence length → 512 or 1024
Keep Gradient Checkpoint ON
Close other GPU-heavy apps (browsers with many tabs, video, other ML tasks)
Then restart training
On the app side, the next patch will detect this specific error and surface the guidance above directly in Smart Alerts, so you don't have to dig through logs. Please let me know if the steps above fix it for you.

**Comment by crazyherozk** [comment]
@tuwenbo0120 thanks

**Comment by crazyherozk** [comment]
@tuwenbo0120 
I've set the power to High Performance and downgraded the display to 60Hz. The training ran for about 300 steps this time, but it still crashed afterward with the same error message.

Max sequence length is 1024，number of lord layers is 16，batch size is 1.

**Comment by tuwenbo0120** [comment]
@crazyherozk 
Thanks for the follow-up — I checked this more carefully and this is **not a RAM / machine size problem** on your side.

Your 96 GB M2 Max should be able to handle this class of workload. The real issue is a confirmed **upstream MLX + macOS Tahoe bug**:

- `ml-explore/mlx#3267`: LoRA training can be killed by the macOS Metal watchdog while the display is active, with the exact same `kIOGPUCommandBufferCallbackErrorImpactingInteractivity` error
- `ml-explore/mlx#3302`: confirms `AGX_RELAX_CDM_CTXSTORE_TIMEOUT=1` as a working workaround

This is why lowering the config helped only partially but did not fully solve it.

I have shipped a fix in **M-Courtyard v0.5.6**:
- the training subprocess now automatically sets `AGX_RELAX_CDM_CTXSTORE_TIMEOUT=1`
- Smart Alerts now recognize this Metal watchdog error and show clearer guidance

Please update to **v0.5.6** and try again:
https://github.com/Mcourtyard/m-courtyard/releases/latest

If it still happens on Tahoe after upgrading, please try this fallback:
- keep the app running
- start training
- then close the MacBook lid so the display is off while the process continues

That workaround is also consistent with the upstream MLX report, because WindowServer stops competing for the GPU when the display is inactive.

Thanks again — your report helped confirm that this was the upstream Tahoe / MLX watchdog issue rather than a model-size limitation.
